pkg/virt-operator/resource/apply/delete.go

```diff
@@ -461,3 +449,126 @@ func DeleteAll(kv *v1.KubeVirt,
 	}
 	return nil
 }
+
+func crdInstanceDeletionCompleted(crd *extv1.CustomResourceDefinition) bool {
+	// Below is an example of what is being looked for here.
+	// The CRD will have this condition once a CRD which is being
+	// deleted has all instances removed related to this CRD.
+	//
+	//    message: removed all instances
+	//    reason: InstanceDeletionCompleted
+	//    status: "False"
+	//    type: Terminating
+
+	if crd.DeletionTimestamp == nil {
+		return false
+	}
+
+	for _, condition := range crd.Status.Conditions {
+		if condition.Type == extv1.Terminating &&
+			condition.Status == extv1.ConditionFalse &&
+			condition.Reason == "InstanceDeletionCompleted" {
+			return true
+		}
+	}
+	return false
+}
+
+func crdFilterNeedFinalizerAdded(crds []*extv1.CustomResourceDefinition) []*extv1.CustomResourceDefinition {
+	filtered := []*extv1.CustomResourceDefinition{}
+
+	for _, crd := range crds {
+		if crd.DeletionTimestamp == nil && !controller.HasFinalizer(crd, v1.VirtOperatorComponentFinalizer) {
+			filtered = append(filtered, crd)
+		}
+	}
+
+	return filtered
+}
+
+func crdFilterNeedDeletion(crds []*extv1.CustomResourceDefinition) []*extv1.CustomResourceDefinition {
+	filtered := []*extv1.CustomResourceDefinition{}
+
+	for _, crd := range crds {
+		if crd.DeletionTimestamp == nil {
+			filtered = append(filtered, crd)
+		}
+	}
+	return filtered
+}
+
+func crdFilterNeedFinalizerRemoved(crds []*extv1.CustomResourceDefinition) []*extv1.CustomResourceDefinition {
+	filtered := []*extv1.CustomResourceDefinition{}
+	for _, crd := range crds {
+		if !crdInstanceDeletionCompleted(crd) {
+			// All crds must have all crs removed before any CRD finalizer can be removed
+			return []*extv1.CustomResourceDefinition{}
+		} else if controller.HasFinalizer(crd, v1.VirtOperatorComponentFinalizer) {
+			filtered = append(filtered, crd)
+		}
+	}
+	return filtered
+}
+
+func crdHandleDeletion(kvkey string,
+	stores util.Stores,
+	clientset kubecli.KubevirtClient,
+	expectations *util.Expectations) error {
+
+	ext := clientset.ExtensionsClient()
+	objects := stores.CrdCache.List()
+
+	crds := []*extv1.CustomResourceDefinition{}
+	for _, obj := range objects {
+		crd, ok := obj.(*extv1.CustomResourceDefinition)
+		if !ok {
+			log.Log.Errorf(castFailedFmt, obj)
+			return nil
+		}
+		crds = append(crds, crd)
+	}
+
+	needFinalizerAdded := crdFilterNeedFinalizerAdded(crds)
+	needDeletion := crdFilterNeedDeletion(crds)
+	needFinalizerRemoved := crdFilterNeedFinalizerRemoved(crds)
+
+	for _, crd := range needFinalizerAdded {
```

Comment:
i was considering the operator mismatch possibility here. I didn't want to run the risk of putting finalizers at the install phase in case someone was using an operator that didn't know about the finializers.

Reply:
I think we can leave it if you insist, but note that in general we break older operators regularly (like when adding a new k8s objects which the old operator does not understand), so not sure if it is worth the extra complication.

Reply:
i think i'd prefer to keep it like it is just so i don't have to think about the scenario where someone's install is stuck due to this finalizer not being removed by an older operator.

Reply:
:+1: